### PR TITLE
Add hit degradation + animation based digging/removing

### PR DESCRIPTION
**Pull request by nomisum**
SP tested, MP test pending

### CODE REVIEW

File: addons/functions/functions/fnc_hitEH.sqf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            [vectorDir _trench, ASLToAGL _position] remoteExec [QFUNC(hitFX), [0,-2] select isDedicated];
+
+            private _progress = _trench getVariable ["ace_trenches_progress", 0];
+            private _multiplier = [configFile >> "CfgVehicles" >> typeOf _trench >> "grad_trenches_damageMultiplier", "NUMBER", 1] call CBA_fnc_getConfigEntry; 
+            private _damage = (_splashDamage*_multiplier)/20; // 1 HE shell appr 25% decay depending on ammo type
+            _progress = _progress - _damage;
+
+            if (_progress > 0) then {
+                private _lift = linearConversion [0, 1, _progress, -([configFile >> "CfgVehicles" >> typeOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry), 0, true];
+                _trench animateSource ["rise", _lift, true];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitEH.sqf:53]: can you call FUNC(setTrenchProgress) here to shorten & deduplicate code a bit

File: addons/functions/functions/fnc_setTrenchProgress.sqf [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    [0, 1, 
+    _progress, 
+    -_offset, 
+    0,
+    true
+    ]);
+
+_trench animateSource ["rise", _lift, true];
+
+diag_log format ["_lift %1, _offset %2, _position %3, _progress %4", _lift, _offset, getpos _trench, _progress];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_setTrenchProgress.sqf:33]: diag_log should not be in production code. 
use TRACE_4() or something

File: addons/functions/functions/fnc_hitEH.sqf [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+if (_trench getVariable [QGVAR(hitHandler), false]) exitWith { 
+    TRACE_1("hitEH","not adding hit handler, already there");
+};
+
+_trench setVariable [QGVAR(hitHandler), true, true];
+
+
+
+_trench addEventHandler ["HitPart", {
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitEH.sqf:29]: hmm this handler is only added on server. [the docs say](https://community.bistudio.com/wiki/Arma_3:_Event_Handlers#HitPart):
> While you can add "HitPart" handler to a remote unit, the respective addEventHandler command must be executed on the shooter's PC and will only fire on shooter's PC as well.

Sooo… this function should be called on the server AND every client I guess?

File: addons/functions/functions/fnc_hitEH.sqf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    (_this select 0) params ["_trench", "_shooter", "_projectile", "_position", "_velocity", "_selection", "_ammo", "_vector", "_radius", "_surfaceType", "_isDirect"];
+
+    diag_log str _isDirect;
+
+    diag_log str (_ammo select 2);
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitEH.sqf:34]: diag_log bad

File: addons/functions/functions/fnc_hitEH.sqf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+            private _progress = _trench getVariable ["ace_trenches_progress", 0];
+            private _multiplier = [configFile >> "CfgVehicles" >> typeOf _trench >> "grad_trenches_damageMultiplier", "NUMBER", 1] call CBA_fnc_getConfigEntry; 
+            private _damage = (_splashDamage*_multiplier)/20; // 1 HE shell appr 25% decay depending on ammo type
+            _progress = _progress - _damage;
+
+            if (_progress > 0) then {
+                private _lift = linearConversion [0, 1, _progress, -([configFile >> "CfgVehicles" >> typeOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry), 0, true];
+                _trench animateSource ["rise", _lift, true];
+                _trench setVariable ["ace_trenches_progress", _progress, true];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitEH.sqf:54]: I smell a race condition here.

Imagine I'm digging a trench. The trench is local to my client (I guess? Or is it the server?). Every 100ms the trench is raised by a bit.

Now you, a different client, shoot up the trench. You're setting the trench progress here (okay, this handler runs on the server only, but – see above reference to the HitPart docs – we'll probably need to add it on every client).

* My client – in the digging PFH – gets the current progress (lets say 0.7)
* Your client's "set progress to 0.4 because I shot it" value arrives
* My client – in the digging PFH – sets the progress it got plus the new progress (0.7+0.05 or so)
-> your damage goes unnoticed; maybe the trench yanks down and up again)

Comment (Fusselwurm) [reply addons/functions/functions/fnc_hitEH.sqf:54]: Suggestion:
animateSource and ace_trenches_progress should ideally be *set AND read* on one node only (client or server, whatever), ideally where the trench is local.

Everything else should work via events, and the node where the trench is local can then make a good decision what to do about them and keep things consistent

File: addons/functions/functions/fnc_hitEH.sqf [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            } else {
+                deleteVehicle _trench;
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitEH.sqf:56]: FUNC(removeTrench)  tells me that removing a trench is a bit more complicated than just a single `deleteVehicle` call

File: addons/functions/functions/fnc_hitEH.sqf (hunk 5 above)
Comment (nomisum) [reply addons/functions/functions/fnc_hitEH.sqf:54]: yes, locality should be fixed for the trench after placing. its not necessary to have fluent building anymore with animations i hope. before it was not possible due to significant lags of setpos.

**Comment by nomisum** [comment]
I am working on making trenches/progress local to the server only. Local only ends on placeConfirm.

File: addons/functions/functions/fnc_hitEH.sqf (hunk 5 above)
Comment (nomisum) [reply addons/functions/functions/fnc_hitEH.sqf]: Locality is changed now, might still need some clean up

File: addons/functions/functions/fnc_spawnTrench.sqf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+* Public: No
+*/
+
+params ["_trenchClass", "_posDiff", "_pos", "_vecDirAndUp"];
+
+//Create a new trench, that is globally visible
+private _trench = createVehicle [_trenchClass, [0,0,0], [], 0, "CAN_COLLIDE"];
+private _digTime = missionNamespace getVariable [getText (configFile >> "CfgVehicles" >> _trenchClass >>"ace_trenches_diggingDuration"), 20];
+
+_trench setObjectTextureGlobal [0, surfaceTexture _pos];
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_spawnTrench.sqf:27]: This should be below the last set, or you will get the Texture from underwater.

File: addons/functions/functions/fnc_initTrench.sqf [hunk 8]
```diff
@@ -25,11 +25,14 @@ if (is3DEN) exitWith {
...
 if (isServer) then {
    _object setVariable ["ace_trenches_progress", 1, true];
-   _object setVariable ["ace_trenches_placeData", [_pos, _vecDirAndUp], true];
 
     if (GVAR(allowTrenchDecay)) then {
        [_object, GVAR(timeoutToDecay), GVAR(decayTime)] call FUNC(decayPFH);
     };
+
+    if (GVAR(allowHitDecay)) then {
+       [_object, GVAR(hitDecayMultiplier)] remoteExec [QFUNC(hitEH), 0, true];
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_initTrench.sqf:34]: Could we use the CBA System instead of remoteExec so that the code works when remoteExec is blocked.

File: addons/functions/functions/fnc_spawnTrench.sqf (hunk 7 above)
Comment (nomisum) [reply addons/functions/functions/fnc_spawnTrench.sqf:27]: not in this case as _pos is target pos not current pos

Comment (AdlerSalbei) [reply addons/functions/functions/fnc_spawnTrench.sqf:27]: We should still move it down to avoide future confusions.

Comment (nomisum) [reply addons/functions/functions/fnc_spawnTrench.sqf:27]: ok i dont mind

File: addons/functions/CfgVehicles.hpp [hunk 9]
```diff
@@ -138,6 +138,7 @@ class CfgVehicles {
         ace_trenches_removalDuration = QGVAR(smallEnvelopeRemovalTime);
         scopecurator = 2;
         scope = 2;
+        grad_trenches_damageMultiplier = 3;
```
Comment (AdlerSalbei) [inline addons/functions/CfgVehicles.hpp:141]: Wouldn't it make sense to make this an Option in the CBA Options?

File: addons/functions/functions/fnc_applyFatigue.sqf [hunk 10]
```diff
@@ -0,0 +1,33 @@
+#include "script_component.hpp"
+/*
+ * Author:  nomisum
+ * Apply Fatigue on Client
+ *
+ * Arguments:
+ * NONE
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_applyFatigue.sqf:7]: Parameters are missing

File: addons/functions/functions/fnc_applyFatigue.sqf [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Return Value:
+ * NONE
+ *
+ * Example:
+ * [] call grad_trenches_functions_fnc_applyFatigue
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_applyFatigue.sqf:13]: Parameters are missing

File: addons/functions/functions/fnc_continueDiggingTrench.sqf [hunk 12]
```diff
@@ -103,7 +94,7 @@ if (_actualProgress == 0) then {
 
 [{
     params ["_args", "_handle"];
-    _args params ["_trench", "_unit", "_digTime", "_vecDirAndUp"];
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_continueDiggingTrench.sqf:106]: Warum wird der Vector hier entfernt?

File: addons/functions/functions/fnc_hitPart.sqf [hunk 13]
```diff
@@ -0,0 +1,30 @@
+#include "script_component.hpp"
+/*
+ * Author: nomisum
+ * Server receives hitPart Events from Clients and calulates
+ *
+ * Arguments:
+ * NONE
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_hitPart.sqf:7]: Parameters are missing

File: addons/functions/functions/fnc_hitPart.sqf [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * Return Value:
+ * NONE
+ *
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_hitPart.sqf:11]: Example is missing

**Comment by AdlerSalbei** [comment]
Nice, mit den Sounds

File: addons/functions/CfgVehicles.hpp (hunk 9 above)
Comment (nomisum) [reply addons/functions/CfgVehicles.hpp:141]: you mean for each trench? its meant to control individual "structural strength" of smaller vs larger trenches. e.g. small trenches are destroyed by one HE-shell, larger only partly.

File: addons/functions/functions/fnc_continueDiggingTrench.sqf (hunk 12 above)
Comment (nomisum) [reply addons/functions/functions/fnc_continueDiggingTrench.sqf:106]: vector wird nur noch einmal gesetzt (d.h. beim platzieren), danach dank animations nicht mehr nötig.

Comment (AdlerSalbei) [reply addons/functions/functions/fnc_continueDiggingTrench.sqf:106]: Guter punkt!

File: addons/functions/CfgVehicles.hpp (hunk 9 above)
Comment (AdlerSalbei) [reply addons/functions/CfgVehicles.hpp:141]: Yeah, so that mission builders can choose.

Comment (nomisum) [reply addons/functions/CfgVehicles.hpp:141]: soo.. what are you suggesting exactly? making a slider for each trench type? disabling the difference between trenches by checkmark?

Comment (AdlerSalbei) [reply addons/functions/CfgVehicles.hpp:141]: No, like a Slider, and then the code in the config would look like this:
`ace_trenches_diggingDuration = QGVAR(giantEnvelopeDigTime);`

File: addons/functions/functions/fnc_spawnTrench.sqf [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+*/
+
+params ["_trenchClass", "_posDiff", "_pos", "_vecDirAndUp"];
+
+//Create a new trench, that is globally visible
+private _trench = createVehicle [_trenchClass, [0,0,0], [], 0, "CAN_COLLIDE"];
+private _digTime = missionNamespace getVariable [getText (configFile >> "CfgVehicles" >> _trenchClass >>"ace_trenches_diggingDuration"), 20];
+
+_trench setVariable [QGVAR(diggingSteps), (_posDiff/(_digTime*10)), true];
+if (GVAR(createTrenchMarker)) then {[_trench, side group _unit] call FUNC(createTrenchMarker)};
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_spawnTrench.sqf:28]: nit pick: `side` can take a unit as parameter, no need to go via `group`.

Uhm. Speaking of which… `_unit` is not defined here, is it :rofl:

File: addons/functions/functions/fnc_setTrenchProgress.sqf [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    true
+    ]);
+
+if (_duration > 0) then {
+    _trench animateSource ["rise", _lift, _duration];
+} else {
+    _trench animateSource ["rise", _lift, true];
+};
+
+TRACE_4("_lift, _offset, _position, _progress", _lift, _offset, getpos _trench, _progress);
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_setTrenchProgress.sqf:37]: this doesnt work, placeholders are missing. 
```suggestion
TRACE_4("_lift: %1, _offset: %2, _position: %3, _progress: %4", _lift, _offset, getpos _trench, _progress);
```

File: addons/functions/functions/fnc_setTrenchProgress.sqf [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+* Return Value:
+* None
+*
+* Example:
+* [TrenchObj, 0.5] call grad_trenches_functions_fnc_setTrenchProgress
+*
+* Public: No
+*/
+
+params ["_trench", "_progress", ["_duration", 0]];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_setTrenchProgress.sqf:19]: in the doc block above the `params`, the third parameter ("_duration") documentation is missing. Its not "always instant" anymore.

This happens easily if things are defined in two places, like in documentation block *and* in params block.

We can combine both a bit more by moving the parameter type documentation into the params block:

```suggestion
params [
    ["_trench", objNull, [objNull]],
    ["_progress", 1.0, [0]],  //  float between 0 and 1
    ["_duration", 0, [0]]
];
```

File: addons/functions/functions/fnc_removeTrench.sqf [hunk 18]
```diff
@@ -118,28 +96,25 @@ private _fnc_condition = {
...
     };
 
-    private _animationPhase = _trench animationSourcePhase "rise";
-    private _diff = (_trench getVariable [QGVAR(diggingSteps), (([configFile >> "CfgVehicles" >> typeOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry)/(_removeTime*10))]) * _diggerCount;
- 
-    _trench animateSource ["rise", _animationPhase - _diff, true];
-    _trench setVectorDirAndUp _vecDirAndUp;
+    private _diggerCount = count (_trench getVariable [QGVAR(diggers),[]]);
+    private _newProgress =  _actualProgress - ((1/_removeTime)/10) * _diggerCount;
+    _trench setVariable ["ace_trenches_progress", _newProgress, true];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_removeTrench.sqf:101]: As far as I can see, `ace_trenches_progress` it is only ever set in conjunction with calling `FUNC(setTrenchProgress)`, so it can be moved into the function.

File: addons/functions/functions/fnc_decayPFH.sqf [hunk 19]
```diff
@@ -67,16 +67,7 @@ if (isNil QGVAR(decayPFH)) then {
 						} else {
 							_trench setVariable ["ace_trenches_progress", _progress];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_decayPFH.sqf:68]: as far as I can tell, here is the only instance where its not set globally. intentional?

Comment (nomisum) [reply addons/functions/functions/fnc_decayPFH.sqf:68]: i have no idea how decay works but progress should never be local.

File: addons/functions/functions/fnc_spawnTrench.sqf (hunk 15 above)
Comment (nomisum) [reply addons/functions/functions/fnc_spawnTrench.sqf]: fixed in 9978ad27836f80854cad36f6e083c694291c5c5f

File: addons/functions/functions/fnc_continueDiggingTrench.sqf [hunk 20]
```diff
@@ -120,24 +111,16 @@ if (_actualProgress == 0) then {
 
     private _actualProgress = _trench getVariable ["ace_trenches_progress", 0];
     private _diggerCount = count (_trench getVariable [QGVAR(diggers),[]]);
-    private _animationPhase = _trench animationSourcePhase "rise";
-    private _diff = (_trench getVariable [QGVAR(diggingSteps), (([configFile >> "CfgVehicles" >> typeOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry)/(_digTime*10))]) * _diggerCount;
-    
-    _trench animateSource ["rise", _animationPhase + _diff, true];
-    _trench setVariable ["ace_trenches_progress", _actualProgress + ((1/_digTime)/10) * _diggerCount, true];
+    private _newProgress =  _actualProgress + ((1/_digTime)) * _diggerCount;
+    _trench setVariable ["ace_trenches_progress", _newProgress, true];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_continueDiggingTrench.sqf:115]: this line can now be removed, right

Comment (Fusselwurm) [inline addons/functions/functions/fnc_continueDiggingTrench.sqf:115]: this line can now be removed, right

Comment (nomisum) [reply addons/functions/functions/fnc_continueDiggingTrench.sqf]: 9b92c5f14b310afac79cf8cb3480bbdb3d7a3a2d

File: addons/functions/CfgVehicles.hpp (hunk 9 above)
Comment (nomisum) [reply addons/functions/CfgVehicles.hpp]: i feel this is a bit too detailed for a setting / cant imagine someone has the need for it and its crowding the already extensive cba settings.

a1fea33ef5a2f3eef8bd7c8120984676bdf672fb

File: addons/functions/functions/fnc_hitPart.sqf [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+ * Public: No
+ */
+
+
+params ["_trench", "_position", "_splashDamage"];
+
+// send fx to clients
+[QGVAR(hitFX), [vectorDir _trench, ASLToAGL (_position)]] call CBA_fnc_globalEvent;
+
+private _progress = _trench getVariable ["ace_trenches_progress", 0];
```
Comment (Fusselwurm) [inline addons/functions/functions/fnc_hitPart.sqf:25]: As far as I can see, you mostly read ace_trench_progress to then add or subtract something from it.

So another suggestion for refactor: 

create new `FUNC(addTrenchProgress)` that contains:

```sqf
params ["_trench", "_addedProgress", "_time"];
private _newProgress = (_trench getVariable["ace_trenches_progress", 0]) + _addedProgress;
if (_newProgress > 0) then {
    [_trench, _newProgress, _time] call FUNC(setTrenchProgress);
} else {
    [_trench] call FUNC(deleteTrench);
};
```
and use that whereever progress is changed :thinking: , so the manipulation of ace_trenches_progress is reduced to fewer places. hmmm. optional though, I think there is a lot of other places that would need change as well.

**Review by Fusselwurm — APPROVED** [review]
I cant find any major poo poo right now. 

I did *eyeballs only* code review though, and did *not* run this code!

File: addons/functions/functions/fnc_applyFatigue.sqf [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+ * Example:
+ * [trenchObj, unitObj] call grad_trenches_functions_fnc_applyFatigue
+ *
+ * Public: No
+ */
+
+// called every 1s on builder
+
+params ["_trench", "_unit"];
+
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_applyFatigue.sqf:22]: Ther should be a check here if the unit is local. Just to be sure.

File: addons/functions/functions/fnc_hitFX.sqf [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Public: No
+ */
+
+ params ["_vectorDir", "_position"];
+
+private _vectorDir = _vectorDir vectorMultiply 1;
+private _ps1 = "#particlesource" createVehicleLocal [0,0,0]; 
+_ps1 setPos _position;
+_ps1 setParticleParams [
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_hitFX.sqf:24]: Could the formating of the Params Array be opptimised, so its easier to read?

File: addons/functions/functions/fnc_applyFatigue.sqf (hunk 22 above)
Comment (nomisum) [reply addons/functions/functions/fnc_applyFatigue.sqf:22]: this is called by a target event – it cant be not-local. 

the worst that can happen is the server fatigue gets drawn out as this is called from server. not sure if the server has stamina though :D

File: addons/functions/functions/fnc_spawnTrench.sqf [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+* 1: PosDiff <NUMBER 0-1>
+* 2: Pos <ARRAY>
+* 3: VecDirAndUp <NESTED ARRAY>
+* 4: Unit <OBJECT>
+*
+* Return Value:
+* None
+*
+* Example:
+* [TrenchObj, 0.5] call grad_trenches_functions_fnc_setTrenchProgress
```
Comment (AdlerSalbei) [inline addons/functions/functions/fnc_spawnTrench.sqf:17]:
```suggestion
* [TypeOf TrenchObj, 0.5, [0,0,0], [[0,0,0], [0,0,0]], ace_player] call grad_trenches_functions_fnc_spawnTrench
```

File: addons/functions/stringtable.xml [hunk 25]
```diff
@@ -533,5 +533,25 @@
             <English>Decay Setting</English>
             <German>Verfall Einstellungen</German>
         </Key>
+        <Key ID="STR_GRAD_Trenches_functions_hitDecaySubCategory">
+            <English>Hit Decay</English>
+            <German>Verfall bei Treffer</German>
```
Comment (AdlerSalbei) [inline addons/functions/stringtable.xml:538]: Würde hier Vorschlagen, Treffer Verfall zu nehmen. Ist kürzer und passt besser zur Englischen Version.
```suggestion
            <German>Treffer Verfall</German>
```

File: addons/functions/functions/fnc_applyFatigue.sqf (hunk 22 above)
Comment (AdlerSalbei) [reply addons/functions/functions/fnc_applyFatigue.sqf:22]: Hm, just thinking about how to make it PEBKAC proof

Comment (nomisum) [reply addons/functions/functions/fnc_applyFatigue.sqf:22]: if that is the goal we need null checks for each and everything too. not sure if this is the quality level we aim for 😁 would bloat up the project quite a bit.

File: addons/functions/stringtable.xml (hunk 25 above)
Comment (nomisum) [reply addons/functions/stringtable.xml:538]: Ist Kürze hier das Kriterium? Ich finds schwieriger zu verstehen, was es meint - würd tatsächlich eher das Englische anpassen, um es eindeutig zu machen.

Comment (AdlerSalbei) [reply addons/functions/stringtable.xml:538]: Muss ehrlich sagen, es ist eigentlich egal.

File: addons/functions/functions/fnc_applyFatigue.sqf (hunk 22 above)
Comment (AdlerSalbei) [reply addons/functions/functions/fnc_applyFatigue.sqf:22]: Thats true.

**Comment by nomisum** [comment]
finished everything i wanted to do, MP test pending

File: addons/functions/functions/fnc_digFX.sqf [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+    case "GRAD_envelope_short" : {  _count = 7; }; 
+    case "ACE_envelope_small" : {  _count = 5; }; 
+    case "ACE_envelope_big" : {  _count = 7; }; 
+    case "GRAD_envelope_giant" : {  _count = 11; }; 
+    case "GRAD_envelope_vehicle" : {  _count = 13; }; 
+    case "GRAD_envelope_long" : {  _count = 15; }; 
+    default {  }; 
+};
+
+for "_i" from 1 to _count do {
```
Comment (DerZade) [inline addons/functions/functions/fnc_digFX.sqf:32]: Instead of hard coding values why not dynamically finding the number of selections. 

You could just set a value within the vehicles config to return how many fx selections there are or find the selections in the script:
```sqf
private _selections = (_vehicle selectionNames "Memory") select {(_x find "fxSource") isEqualTo 0};
```
And then iterate the `_selections` with a `forEach` loop.

**Review by DerZade — CHANGES_REQUESTED** [review]

File: addons/functions/functions/fnc_digFX.sqf (hunk 26 above)
Comment (nomisum) [reply addons/functions/functions/fnc_digFX.sqf:32]: i was about to dynamically detect and then i loaded the wrong mod and didnt find the selectionnames and then i did the other solution to workaround my non existing limitation

Comment (nomisum) [reply addons/functions/functions/fnc_digFX.sqf]: hätte `isNotEqualTo` sein müssen :reeeee:

Comment (DerZade) [reply addons/functions/functions/fnc_digFX.sqf]: Whut? Hold up!!

The [find Alternative Syntax](https://community.bistudio.com/wiki/find#Alternative_Syntax) docs state:
> Zero based position of the first sequence of characters that matches substring, -1 if not found

You want to find all selection names, which start `"fxSource"`. Correct?

So the expression `_name find "fxSource"` should return `0` for the items you want to keep -> You use `(_x find "fxSource") isEqualTo 0` ( NOT `isNotEqualTo`) as your condition, because you want to keep all selection names, which start with `"fxSource"`.

Am I getting something wrong here?

Comment (nomisum) [reply addons/functions/functions/fnc_digFX.sqf]: idk i didnt even think about it but it didnt work and reversing did

Comment (DerZade) [reply addons/functions/functions/fnc_digFX.sqf]: image

Comment (DerZade) [reply addons/functions/functions/fnc_digFX.sqf]: I'm pretty sure we want the first one 😅

File: addons/functions/functions/fnc_deleteTrench.sqf [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * Public: No
+ */
+
+params ["_trench"];
+
+_trench setVariable ["ace_trenches_digging", false, true];
+_trench setVariable [QGVAR(diggingType), nil, true];
+_unit setVariable [QGVAR(diggingTrench), false, true];
+[QGVAR(addDigger), [_trench, _unit, false, true]] call CBA_fnc_serverEvent;
```
Comment (DerZade) [inline addons/functions/functions/fnc_deleteTrench.sqf:23]: Where the fuck does `_unit` come from? And why are we addingDiggers in `fnc_deleteTrench`? 😅

File: addons/functions/functions/fnc_digFX.sqf (hunk 26 above)
Comment (nomisum) [reply addons/functions/functions/fnc_digFX.sqf]: test with the real thing:

Comment (nomisum) [reply addons/functions/functions/fnc_digFX.sqf]: seems like a case sensitive issue

File: addons/functions/functions/fnc_deleteTrench.sqf (hunk 27 above)
Comment (nomisum) [reply addons/functions/functions/fnc_deleteTrench.sqf]: handleDiggerToGVAR seems to be lost/renamed somewhen into addDigger which is ... wrong

**Comment by nomisum** [comment]
- [x] dedicated test for myself
- [ ] add/removeDigger with others

File: addons/functions/initSettings.sqf [hunk 28]
```diff
@@ -10,24 +10,35 @@
...
+[QGVAR(vehicleEnvelopeRemovalTime), "SLIDER", localize LSTRING(VehicleEnvelopeRemovalTime), [localize LSTRING(settingCategory), localize LSTRING(trenchSubCategory)], [5, 1200, 120, 0], true] call CBA_fnc_addSetting;
 [QGVAR(allowLongEnvelope), "CHECKBOX", [localize LSTRING(allowLongEnvelope_displayName), localize LSTRING(allowLongEnvelope_tooltip)], [localize LSTRING(settingCategory), localize LSTRING(trenchSubCategory)], true, true] call CBA_fnc_addSetting;
 [QGVAR(LongEnvelopeDigTime), "SLIDER", localize LSTRING(LongEnvelopeDigTime), [localize LSTRING(settingCategory), localize LSTRING(trenchSubCategory)], [5, 1200, 100, 0], true] call CBA_fnc_addSetting;
-[QGVAR(LongEnvelopeRemovalTime), "SLIDER", localize LSTRING(LongEnvelopeRemovalTime), [localize LSTRING(settingCategory), localize LSTRING(trenchSubCategory)], [-1, 1200, -1, 0], true] call CBA_fnc_addSetting;
+[QGVAR(LongEnvelopeRemovalTime), "SLIDER", localize LSTRING(LongEnvelopeRemovalTime), [localize LSTRING(settingCategory), localize LSTRING(trenchSubCategory)], [5, 1200, 100, 0], true] call CBA_fnc_addSetting;
```
Comment (AdlerSalbei) [inline addons/functions/initSettings.sqf:28]: Das war schon im code berücksichtigt, du fügst damit nur unnötige doppelarbeit hinzu und entfernst backword compatability. Siehe hier:
https://github.com/gruppe-adler/grad_trenches/blob/master/addons/functions/functions/fnc_removeTrench.sqf#L38

Comment (nomisum) [reply addons/functions/initSettings.sqf:28]: eli5, raff ich nicht. backward zu was. doppelarbeit für wen.

wenn du sagst presets sollen unterschiedlich sein für dig/removal dann lassma die so 🤔

Comment (AdlerSalbei) [reply addons/functions/initSettings.sqf:28]: Wenn du mal in den Code schaust, dann ist da ganz klar drin, das -1 bedeutet das es den gleichen wert wie dig hat.

Wenn du jetzt also den Code dazu änderst das es -1 nicht mehr gibt, dann bricht das alle alten missionen.
Und alle änderungen die gruppen an der Dig Time gemacht haben, müssen diese jetzt auch auf die removal time übertragen.

**Review by DerZade — APPROVED** [review]
Haven't tested ingame, this is just a review regarding code quality.
